src/Controller/Backend/Invitation.php

```diff
@@ -0,0 +1,305 @@
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\Extension\Core\Type\TextareaType;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
+ * Backend controller for invitation code generation.
+ *
+ * @author Carlos Perez <mrcarlosdev@gmail.com>
+ */
+class Invitation extends BackendBase
+{
+    protected function addRoutes(ControllerCollection $c)
+    {
+        $c->match('/invitationlink', 'invitationLink')
+            ->bind('invitationlink');
+
+        $c->match('/generatelink', 'generateLink')
+            ->bind('generatelink');
+
+        $c->match('/sendlink', 'sendLink')
+            ->bind('sendlink');
+
+    }
+
+    /**
+     * Invitation link route.
+     *
+     * @return \Bolt\Response\BoltResponse|\Symfony\Component\HttpFoundation\RedirectResponse
+     */
+    public function invitationLink()
+    {
+        // Get and generate the base form to generate the invitation code
+        $form = $this->getGenerateInvitationForm($this->getUser());
+        $form = $form->getForm();
+
+        /** @var \Symfony\Component\Form\FormView|\Symfony\Component\Form\FormView[] $formView */
+        $formCodeView = $form->createView();
+
+        // Get and generate the base form to share the invitation code by email
+        $form = $this->getInvitationEmailForm($this->getUser());
+        $form = $form->getForm();
+
+        /** @var \Symfony\Component\Form\FormView|\Symfony\Component\Form\FormView[] $formView */
+        $formEmailView = $form->createView();
+
+        // Get the current user to know what user role they can invite
+        $currentUser = $this->getUser();
+
+        $manipulatableRoles = $this->app['permissions']->getManipulatableRoles($currentUser->toArray());
+        foreach ($formCodeView['roles'] as $role) {
+            if (!in_array($role->vars['value'], $manipulatableRoles)) {
+                $role->vars['attr']['disabled'] = 'disabled';
+            }
+        }
+
+        // Preparing the forms for the view
+        $context = [
+            'form' => $formCodeView,
+            'emailform' => $formEmailView,
+        ];
+
+        return $this->render('@bolt/invitation/generate.twig', $context);
+    }
+
+    /**
+     * Generate invitation code link.
+     *
+     * @param Request $request The Symfony Request
+     *
+     * @return \Bolt\Response\BoltResponse|\Symfony\Component\HttpFoundation\RedirectResponse
+     */
+    public function generateLink(Request $request)
+    {
+        // Get the Roles for the invitation code
+        $roles = [];
+
+        if (null !== ($request->request->get('roles'))) {
+            $roles = $request->request->get('roles');
+        }
+
+        // Get the expiration date for the invitation code
+        $expiration_date = $request->request->get('expiration_date');
+        $expiration_time = $request->request->get('expiration_time');
```

Comment:
Re-read my comment and thought it might lack context…

My thinking is that the expiration time should be a policy decision, and the person entrusted with issuing these invitations might not have policy delegation rights. More-so, having invitations that are valid (and un-monitored under normal use) with long expiration dates are a **big** security no-no.